nussl/separation/nmf_mfcc.py

```diff
@@ -0,0 +1,215 @@
+            self.clusterer = sklearn.cluster.KMeans(n_clusters=self.num_sources, random_state=self.kmeans_random_seed,
+                                                    **kmeans_kwargs)
+        else:
+            self.clusterer = sklearn.cluster.KMeans(n_clusters=self.num_sources, random_state=self.kmeans_random_seed)
+
+    def run(self):
+        """ This function calls TransformerNMF on the magnitude spectrogram of each channel in the input audio signal.
+        The templates and activation matrices returned are clustered using K-Means clustering. These clusters are used
+        to create mask objects for each source. Note: The masks are not returned in a particular order, but they are in
+        the same order for each channel.
+
+        Returns:
+            self.masks (np.array): A list of binary mask objects that can be used to extract the sources.
+
+        Example:
+
+        .. code-block:: python
+            :linenos:
+
+            signal = nussl.AudioSignal(path_to_input_file='input_name.wav')
+
+            # Set up and run NMF MFCC
+            nmf_mfcc =  nussl.NMF_MFCC(signal, num_sources=2) # Returns a binary mask by default
+            masks = nmf_mfcc.run()
+
+            # Get audio signals
+            sources = nmf_mfcc.make_audio_signals()
+
+            # Output the sources
+            for i, source in enumerate(sources):
+                output_file_name = str(i) + '.wav'
+                source.write_audio_to_file(output_file_name)
+        """
+        self.masks = []
+        self.input_audio_signal.stft_params = self.stft_params
+        self.signal_stft = self.input_audio_signal.stft()
+
+        for i in range(self.input_audio_signal.num_channels):
+            channel_stft = self.input_audio_signal.get_magnitude_spectrogram_channel(i)
+
+            # Set up NMF and run
+            nmf = transformer_nmf.TransformerNMF(input_matrix=channel_stft, num_components=self.num_templates,
+                                                 seed=self.random_seed)
+
+            # Pass these to constructor (above)
+            nmf.should_use_epsilon = False
+            nmf.max_num_iterations = self.num_iterations
+            nmf.distance_measure = self.distance_measure
+            channel_activation_matrix, channel_templates_matrix = nmf.transform()
+
+            # Cluster the templates matrix into Mel frequencies and retrieve labels
+            cluster_templates = librosa.feature.mfcc(S=channel_templates_matrix,
+                                                     n_mfcc=self.n_mfcc)[self.mfcc_start:self.mfcc_end]
+            self.clusterer.fit_transform(cluster_templates.T)
+            self.labeled_templates = self.clusterer.labels_
+
+            # Extract sources from signal
+            channel_masks = self._extract_masks(channel_stft, channel_templates_matrix, channel_activation_matrix)
+            self.masks.append(channel_masks)
+
+        return self.masks
```

Comment:
Okay, so `self.masks` (which will be changed to `self.result_masks`) is a list of lists, no? The documentation needs to reflect that.